### ISSUE TITLE
[prodc-integration] Normal Sleep Power Consumption High

### ISSUE THREAD

**Issue by jettr** [lowRISC/opentitan#22241]
### Description

We are noticing that the power draw while the chip is in normal sleep is around 13mW (3.95mA @ 3.3V). We are hoping to target the 6 - 8 mW range or better for normal sleep.

By normal sleep, I mean a sleep that retains SRAM contents and will wake up and start executing code from where it left off. Normal sleep is like a deeper WFI where there is no pending interrupt to process. We still want to wake up from normal sleep in a reasonable amount of time to handle incoming TPM communication from the AP (either on I2C or SPI bus).

What we are doing for normal sleep:
 * Following [programmers guide][0] to enter low power mode
   *   We configure the [`control`] to have only `MAIN_PD_N` set (this gives us 3.95mA draw)
   *   If we set `MAIN_PD_N` + `USB_CLK_EN_LP` + `IO_CLK_EN` + `CORE_CLK_EN` we get 7.8mA as a point of reference.
       * What do we lose by allowing  `USB_CLK_EN_LP`,  `IO_CLK`, and `CORE_CLK` to shutdown during sleep?

What else can we try to get the power consumption of normal sleep down to the 6-8 mW range or better?

---

For another data point: "deep sleep," i.e. sleep that must resume execution from ROM/ROM_EXT, power consumption looks good. It is around 600uW (0.20mA @ 3.33V).

[0]:https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/programmers_guide.html#programmer-sequence-for-entering-low-power
[`control`]:https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control

**Comment by johngt** [comment]
@andreaskurth / @vogelpi - for power utilisation discussions

**Comment by andreaskurth** [comment]
Potentially related: https://github.com/lowRISC/opentitan/issues/22223

**Comment by moidx** [comment]
AI @jettr to disable entropy complex and measure power consumption to triage #22223. See reference: https://github.com/lowRISC/opentitan/blob/603430849d683ad52a5001fda5d81bbfaa437505/sw/device/lib/testing/entropy_testutils.c#L214.

**Comment by jettr** [comment]
Disabling the entropy complex did reduce power consumption in normal sleep! The power consumption drops to 9mA (2.71mA @ 3.3V). I didn't try re-enabling the entropy sources or do crypto in my tests.

I performed

```rust
EDN0.stop();
EDN1.stop();
csrng_stop();
entropy_src_stop();
```

before the normal sleep where those resolve to

```rust
    impl Edn {
        pub fn stop(&self) {
            self.register
                .ctrl
                .modify(CTRL::CMD_FIFO_RST.val(multibit::size_4bit::TRUE));
            self.register.ctrl.set(0x9999);
        }
    }

    pub unsafe fn csrng_stop() {
        StaticRef::new(CSRNG_BASE_ADDR as *const CsrngRegisters).ctrl.set(0x0999);
    }

    pub unsafe fn entropy_src_stop() {
        let register = StaticRef::new(ENTROPY_SRC_BASE_ADDR as *const EntropySrcRegisters);
        register.module_enable.set(0x9);
        register.entropy_control.set(0x99);
        register.conf.set(0x2649999);
        register.health_test_windows.set(0x600200);
        register.alert_threshold.set(0xfffd0002);
    }
```

This is good new and seems like something we can use. Is there any other IP block we can disable to get another 2 - 3 mW savings?

**Comment by vogelpi** [comment]
Thanks for the feedback @jettr ! This is good news. Could please try to only disable the ENTROPY_SRC, i.e., just do a `entropy_src_stop()`? My expectation would be that this alone should be sufficient to reach 9mA. If you use the function `dif_entropy_src_set_enabled()` instead of `entropy_src_stop()` you also re-enable the ENTROPY_SRC with the same settings in an attempt to resume. In contrast, the stop function clears some of the settings.

As for further things to try out: Reading the documentation of the register you referenced above. I would suggest to use the following:
- [MAIN_PD_N](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control--main_pd_n) = 1. IIUC, you enter low power state otherwise, where the SRAM content is not retained. Note there is however the retention SRAM which always keeps its content also in low power mode. It's separate option to explore.
- [CONTROL . USB_CLK_EN_ACTIVE](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control--usb_clk_en_active) = 0
- [CONTROL . USB_CLK_EN_LP](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control--usb_clk_en_lp) = 0
- [CONTROL . IO_CLK_EN](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control--io_clk_en) = 0
- [CONTROL . CORE_CLK_EN](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/pwrmgr/doc/registers.html#control--core_clk_en) = 0

Apart from that, you also have the option to disable individual clocks via clkmgr. However, I don't know how this interacts with above settings.

- [CLK_ENABLES](https://opentitan.org/book/hw/ip/clkmgr/doc/registers.html#clk_enables) allows you to explicitly enable/disable the USB, and the 3 peripheral clocks.
- [CLK_HINTS](https://opentitan.org/book/hw/ip/clkmgr/doc/registers.html#clk_hints) allows you to explicitly enable/disable the clocks of the big crypto accelerators.

Another analog part you may want to disable is the ADC, see [adc_en_ctl](https://opentitan.org/book/hw/ip/adc_ctrl/doc/registers.html#adc_en_ctl).

**Comment by jettr** [comment]
> Could please try to only disable the ENTROPY_SRC, i.e., just do a `entropy_src_stop()`? My expectation would be that this alone should be sufficient to reach 9mW.

You are correct. It appears that the only `entropy_src_stop()` reduces the power consumption.

> If you use the function `dif_entropy_src_set_enabled()` instead of `entropy_src_stop()` you also re-enable the ENTROPY_SRC with the same settings in an attempt to resume.

When I only use `entropy_src_set_enabled` and none of the other methods we have been experimenting with in the following way:

```rust
entropy_src_set_enabled(false);
rv32i::support::wfi();
entropy_src_set_enabled(true);
```

I actually get even lower normal sleep power:  8.8mW (2.66mA * 3.3V), but not by that much. This also seems like it is a better choice we as don't have to go through health checks etc again on normal sleep resume.

As for the PMU Control register: I am already using `MAIN_PD_N` as the only bit that is set when entering normal sleep. And if I remove even that, then the chips goes into deep sleep (which is desirable in different sleep scenarios).

> Apart from that, you also have the option to disable individual clocks via clkmgr

We haven't done too much of that yet. That is something we will explore too. Thank you.

> Another analog part you may want to disable is the ADC

We do need to ADC to be monitoring for a cable insertion during normal/deep sleep so we can't fully disable it, but I do hope that we can make full use of the lower power mode that it offers.

**Comment by andreaskurth** [comment]
Related to high entropy consumption

**Comment by vogelpi** [comment]
@jettr We've found out today that keymgr doesn't behave ideally in terms of entropy consumption (see https://github.com/lowRISC/opentitan/issues/22819). We're now discussing a hardware fix for that. In the meantime, and in particular for the engineering sample, software should set the [reseed interval register](https://opentitan.org/book/hw/ip/keymgr/doc/registers.html#reseed_interval_shadowed) to the maximum value. It's a 16-bit register, so the max reseed interval is something like 65535. It's not ideal but should already be much better than the default 256. You may want to try this out on the chip and re-measure power.

**Comment by jettr** [comment]
Thanks for the heads up! We will want to re-measure power for A1, but it should have this change.

**Comment by andreaskurth** [comment]
@jettr: With the relaxed keymgr reseed configuration, does the power consumption of ES match the expectation / requirements?

**Comment by jettr** [comment]
I saw a small power different when I set [reseed interval register](https://opentitan.org/book/hw/ip/keymgr/doc/registers.html#reseed_interval_shadowed) to `0xFFFF`. The normal sleep current for this round of testing started at ~4.26mA and dropped to ~4.07mA with the change. Looks like only about 630uW of power savings.

**Comment by vogelpi** [comment]
Thanks @jettr for getting back with additional results. It's less than I expected but on the other hand, keymgr is still triggering a reseed (CSRNG operation) every 0.5ms with this new setting. I think there is not much more that can be done for ES on the entropy complex front.

**Comment by jettr** [comment]
That's okay, we will just need to re-measure power with the A1 silicon :)

**Comment by andreaskurth** [comment]
@johngt Please add this to PROD SiVal

**Comment by matutem** [comment]
@jettr did you try disabling the various CONTROL.*CLK_EN?

**Comment by jettr** [comment]
> @jettr did you try disabling the various CONTROL.*CLK_EN?

Yes, I performed all of this normal sleep testing with all of the CONTROL.*CLK_EN bits set to 0. The only enabled bits on the CONTROL register are `CONTROL::LOW_POWER_HINT + CONTROL::MAIN_PD_N`

**Comment by vogelpi** [comment]
Hi @jettr , it has been a while... could you re-confirm these measurements on A1/A2? If yes, can we close this or is there something to consider for Earl Grey 2?

**Comment by jettr** [comment]
I am not at the office for a bit, but I bet @vsukhoml would be able to pull
this sleep number from our automated tests running on the new silicon.

**Comment by vogelpi** [comment]
Thanks @jettr, I've now reached out to @vsukhoml for input.

**Comment by vogelpi** [comment]
I now got feedback from @vsukhoml regarding A1/A2.
